### PR TITLE
Add MeshStandardMaterialSG

### PR DESCRIPTION
**Pull request by takahirox**
#10985

This PR
- adds `MeshStandardMaterialSG` which can handle glTF Specular-Glossiness extension.
- lets `GLTF2Loader` support glTF Specular-Glossiness extension.

You can see `MeshStandardMaterialSG` example in In `webgl_loader_gltf2` example
by selecting Specular-Glossiness extension with BoomBox model.

Only the difference between `MeshStandardMaterial` and `MeshStandardMaterialSG`
is `metalness/roughness` or `specular/glossiness`.
So I made `MeshStandardMaterialCommon` and let them inherit it.

/cc @donmccurdy @WestLangley

### CODE REVIEW

**Comment by takahirox** [comment]
Well, let me close once.
I think I can rename `specular2` with `specular`.
Let me try...

**Comment by takahirox** [comment]
Updated and reopened

**Comment by WestLangley** [comment]
Nice job.

I personally do not care for your refactoring, however. You have taken an implementation detail, and pushed it up to the API and created another layer of obfuscation -- especially for users -- by creating the new `MeshStandardMaterialCommon`, which is sort of like a virtual base class.

As I suggested [here](https://github.com/mrdoob/three.js/issues/10985#issuecomment-285908158), leave `MeshStandardMaterial` as-is, and extend it -- similarly to your implementation of `MeshToonMaterial`.

Maybe something like this:

```javascript 
function MeshStandardMaterialSG( parameters ) {

	MeshStandardMaterial.call( this );

	this.defines[ 'STANDARD_SG' ] = '';

	this.type = 'MeshStandardMaterialSG';

	this.glossiness = 0.5;
	this.specular = new Color( 0x111111 );

	this.glossinessMap = null;

	this.specularMap = null;

	delete this.metalness;
	delete this.roughness;
	delete this.metalnessMap;
	delete this.roughnessMap;

	this.setValues( parameters );

}
```
Yes, I understand that you can argue otherwise from a software-design standpoint. However, I expect that 99% of users will use `MeshStandardMaterial`, and the `SG` version only by sophisticated users who prefer the specular-glossiness workflow.

**Comment by takahirox** [comment]
OK, that makes sense.
I'll update.

File: src/renderers/shaders/ShaderChunk/uv_pars_fragment.glsl [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#if defined( USE_MAP ) || defined( USE_BUMPMAP ) || defined( USE_NORMALMAP ) || defined( USE_SPECULARMAP ) || defined( USE_ALPHAMAP ) || defined( USE_EMISSIVEMAP ) || defined( USE_ROUGHNESSMAP ) || defined( USE_METALNESSMAP )
+#if defined( USE_MAP ) || defined( USE_BUMPMAP ) || defined( USE_NORMALMAP ) || defined( USE_SPECULARMAP ) || defined( USE_ALPHAMAP ) || defined( USE_EMISSIVEMAP ) || defined( USE_ROUGHNESSMAP ) || defined( USE_METALNESSMAP ) || defined( USE_GLOSSINESSMAP) || defined( USE_SPECULAR2MAP )
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderChunk/uv_pars_fragment.glsl]: `SPECULAR2MAP` ?
defined( USE_GLOSSINESSMAP ) // white space

It is unfortunate that `specularMap` is used for multiple purposes. We will have to fix that. It should be RGB and represent the specular reflectance of the material.

File: src/renderers/shaders/ShaderLib.js [hunk 2]
```diff
@@ -82,12 +82,15 @@ var ShaderLib = {
 			UniformsLib.displacementmap,
 			UniformsLib.roughnessmap,
 			UniformsLib.metalnessmap,
+			UniformsLib.glossinessmap,
 			UniformsLib.fog,
 			UniformsLib.lights,
 			{
 				emissive: { value: new Color( 0x000000 ) },
 				roughness: { value: 0.5 },
 				metalness: { value: 0 },
```
Comment (WestLangley) [inline src/renderers/shaders/ShaderLib.js]: Yes, should be

    metalness: { value: 0.5 }

Comment (takahirox) [inline src/renderers/shaders/ShaderLib.js]: That isn't what I edited.
I already noticed that and sent another PR.

https://github.com/mrdoob/three.js/pull/11039

Comment (WestLangley) [inline src/renderers/shaders/ShaderLib.js]: Yes, I saw that PR earlier. :-)

Comment (takahirox) [inline src/renderers/shaders/ShaderLib.js]: Ah, ok. I understand what you meant.

File: src/renderers/shaders/ShaderChunk/uv_pars_fragment.glsl (hunk 1 above)
Comment (takahirox) [inline src/renderers/shaders/ShaderChunk/uv_pars_fragment.glsl]: If I'm right, this is a related comment? https://github.com/mrdoob/three.js/issues/9339#issuecomment-233160823

**Comment by takahirox** [comment]
Updated

**Comment by mrdoob** [comment]
As discussed in the SFHTML5 meetup (😁), I think it'll be better if, for now, `GLTF2Loader` just produces a `ShaderMaterial` for the specular/glossiness case.

**Comment by takahirox** [comment]
Alright, I wanna try that before I leave the USA!

**Comment by WestLangley** [comment]
>As discussed in the SFHTML5 meetup (😁), I think it'll be better if, for now, GLTF2Loader just produces a ShaderMaterial for the specular/glossiness case.

That suits me fine.

I assume we then do not require the addition of `MeshStandardMaterialSG` ?

**Comment by mrdoob** [comment]
> I assume we then do not require the addition of `MeshStandardMaterialSG` ?

Yeah, that's the idea. I personally didn't like the name and I couldn't think of a better name. At this point this is mainly for the GLTF2Loader anyway.

**Comment by takahirox** [comment]
Well, I couldn't make a time.
I'll implement after I land in Japan ;D

**Comment by bhouston** [comment]
The MeshPhong model is specular/glossy.  If we merge this, we should remove MeshPhong if this is accepted.  We really shouldn't have two materials that are basically the same thing.

**Comment by WestLangley** [comment]
>we should remove MeshPhong if this is accepted.

In light of what @bhouston said, I agree. I did not realize that was an option.

Removing `Phong` would make our shader code simpler, plus `Phong`, as implemented, does not model metals well, and the way it handles environment map reflections is hacky.

We could keep `Lambert`, which is non-specular and implements Gouraud shading.

**Comment by takahirox** [comment]
So we have two options.

1. Replacing `MeshPhongMaterial`
2. `ShaderMaterial`

Which one do you folks prefer?

/cc @mrdoob

**Comment by WestLangley** [comment]
In spite of the fact that I am not particularly fond of the name `MeshStandardMaterialSG`, my preference is to merge this, and do what I suggested in my previous comment.

**Comment by bhouston** [comment]
The two parameterizations are equivalent.  Thus in a best case scenario it would be 

```
var mat = new MeshStandardMaterial();
mat.parameterMode = THREE.RoughnessMetallic; // sets a define
mat.roughness = ...;
mat.metallic = ...;

//OR

mat.parameterMode = THREE.SpecularGlossiness;
mat.specular = ...;
mat.glossiness = ...;
```

Because it really is the same material just different parameterization.

**Comment by WestLangley** [comment]
Personally, I think it is confusing to have a single `MeshStandardMaterial` with all these parameters:
```
material.workflow //THREE.RoughnessMetallic or THREE.SpecularGlossiness
material.roughness
material.metalness
material.specular
material.glossiness
material.roughnessMap
material.metalnessMap
material.specularMap
material.glossinessMap
```
Hence, a derived class was implemented -- a class with a lousy name, unfortunately, but a clear API.

I prefer to use this material as a replacement for Phong, too, as @bhouston suggested.

**Comment by mrdoob** [comment]
Merging this as is would require maintenance in the future. I still prefer to just produce custom `ShaderMaterial` in `GLTF2Loader` so we can support specular/glossiness models. Once that's solved, we can discuss how to move the code to core.

**Comment by WestLangley** [comment]
OK, just remember that although creating a custom `ShaderMaterial` that replicates `MeshStandardMaterialSG` is doable, updating the uniforms and setting the defines must be done by the user at the application layer, and for such an intricate model, doing so is a pain in the neck.

**Comment by mrdoob** [comment]
Yeah, hopefully we can use `onBeforeRender()` for that?

**Comment by takahirox** [comment]
I've implemented Specular-Glossiness extension support with `ShaderMaterial`.

On the current implementation, `GLTF2Loader` setups its `uniforms`.
Users need to update its `uniforms` value(ex: `material.uniforms.map.value`) and its `defines`,
not its properties(ex: `material.map`), if they wanna update material values.
(I think this's `ShaderMaterial` standard manner.)

I'll try `onBeforeRender()` later to allow users to update its property.

The code would be like this, 

    object.onBeforeRender = function( renderer, scene, camera, geometry, material, group ) {

        material.uniforms.map.value = material.map;

    }

**Comment by takahirox** [comment]
Done

**Comment by mrdoob** [comment]
Sweeeet!

**Comment by mrdoob** [comment]
@donmccurdy looks good?

File: examples/js/loaders/GLTF2Loader.js [hunk 3]
```diff
@@ -2312,6 +2464,47 @@ THREE.GLTF2Loader = ( function () {
...
+							uniforms.glossiness.value = material.glossiness;
+
+							uniforms.map.value = material.map;
+							uniforms.aoMap.value = material.aoMap;
+							uniforms.bumpMap.value = material.bumpMap;
+							uniforms.emissiveMap.value = material.emissiveMap;
+							uniforms.envMap.value = material.envMap;
+							uniforms.normalMap.value = material.normalMap;
+							uniforms.specularMap.value = material.specularMap;
+							uniforms.glossinessMap.value = material.glossinessMap;
```
Comment (WestLangley) [inline examples/js/loaders/GLTF2Loader.js]: Do you need to add these? Maybe others?
```
uniforms.normalScale
uniforms.displacementMap
uniforms.displacementScale
uniforms.displacementBias
uniforms.envMapIntensity
uniforms.opacity
uniforms.offsetRepeat

```

Comment (takahirox) [inline examples/js/loaders/GLTF2Loader.js]: Oops, right.
I think we should also have them and others
for the `StandardMaterial` compatibility as much as possible.
I'll update...

Comment (WestLangley) [inline examples/js/loaders/GLTF2Loader.js]: Yes, this is a pain, but it is illustrative to show how it is done...

It is also why I prefer the oddly-named `MeshStandardMaterialSG` as a built-in material over this `ShaderMaterial` approach.

**Comment by takahirox** [comment]
Updated

**Comment by WestLangley** [comment]
I know I have been overruled, but `MeshSGStandardMaterial` is sounding even better to me. : - )

**Comment by donmccurdy** [comment]
Nice work @takahirox!

The implementation of this approach looks good to me, and I don't have any comment on `MeshSGStandardMaterial` vs `ShaderMaterial` here.

If possible in a clean way, I might prefer to isolate the spec-gloss code in an extension object. Something like:

```js
class GLTFSpecGlossExtersion {
  extendMaterialParams (params, ...) {
     params.foo = bar;
     // ...
  }

  createShaderMaterial (params) {
    // ...
    material.onBeforeRender = function () { /* ... */ };
    return material;
  }
}
```

This helps readability IMHO, avoiding 500+ line methods, and the division is clean enough because the specification for extensions is self-contained already. But I'm OK with the code as-is, if we prefer to have all material code in one larger method.

**Comment by mrdoob** [comment]
> I know I have been overruled, but `MeshSGStandardMaterial` is sounding even better to me. : - )

Lets wait to see if people actually use this workflow.

**Comment by takahirox** [comment]
Updated!

**Comment by donmccurdy** [comment]
Great thanks — looks good to me!

**Comment by mrdoob** [comment]
Very nice!

**Comment by mrdoob** [comment]
Thanks!

File: examples/js/loaders/GLTF2Loader.js [hunk 4]
```diff
@@ -2475,6 +2464,47 @@ THREE.GLTF2Loader = ( function () {
 
 					}
 
+					// for Specular-Glossiness
+					if ( child.material && child.material.type === 'ShaderMaterial' ) {
```
Comment (WestLangley) [inline examples/js/loaders/GLTF2Loader.js]: @takahirox Can you change this so `material.type` is not used for this purpose? I do not think that pattern is used elsewhere...

In this case, it should be something like
```
if ( child.material && child.material.isSpecularGlossinessMaterial ) {
```
